### ISSUE TITLE
Support for QuerySet values() and values_list() methods

### ISSUE THREAD

**Issue by rwhitman** [django-salesforce/django-salesforce#142]
I noticed that there have been some new releases that add support for Django 1.8 & 1.9 with the continued issue of the queryset values() and values_list() methods being broken, per latest documentation.

I support a django-salesforce dependent project that unfortunately made heavy use of both these methods, and noticed the problem isn't being tracked in an issue ticket that I can follow. So kicking one off here.

Is there a solution in the pipeline for the near term, or does this seem like something that will be ongoing? Any suggestions out there for an efficient workaround with some longevity? Trying to determine best course of action for supporting Django 1.9 in the near-ish future.

I wish I had the time to understand the guts of django-salesforce a little better and help pitch in to patch, but unfortunately the best I can do is file a ticket at the moment...

Also can't express this enough but this software is fantastic. We would not have been able to use Django for this application had it not existed. To all those who contribute to this project, your hard work is recognized and greatly appreciated. Thank you!

**Comment by hynekcer** [comment]
Is your dependent project private or an interesting open source? (for my motivation) I'm working on SOAP and Bulk API used internally for big and huge number of fast insert/update/delete. I'm working on canvas middleware for easy integration of an Django driven controller into SFDC pages, but you are right there is no issue for values() and values_list().

**Comment by rwhitman** [comment]
The project is private unfortunately. But here are some sanitized example lines where we use values_list:

```return list(OpportunityContactRole.objects.filter(contact_id=contact_id).values_list('opportunity_id', flat=True))```

```some_q.add(another_q | Q(pk__in=list(PricebookEntry.objects.filter(another_q).values_list('product2_id', flat=True))), Q.AND)```

**Comment by philchristensen** [comment]
Thanks for opening this issue, and for your support of the project! It's a good idea to use this issue to track the progress towards complete 1.8/1.9 support, but there isn't a specific plan for how we're going to get there.

The initial version of this library required a pretty substantial amount of reverse-engineering of Django, and although I haven't really sat down and compared the two branches, I have a feeling that true support for >1.8 is something of an either-or proposition.

Basically, I'm thinking that 1.8 support may be best accomplished by a rewrite that breaks backwards compatibility, at least initially.

**Comment by hynekcer** [comment]
Django uses `Python Database API Specification v2.0` in drivers. It prefers tuples for query rows returned by `fetch*` instead of dictionaries. See this answer in [DB API FAQ](https://www.python.org/dev/peps/pep-0249/#frequently-asked-questions): case insensitive names in many databases and unexped names of database generated columns like aggregation functions. Django also uses tuples and sometimes adds also dictionary keys, sometimes use only indexes internally. We solved many features directly in ORM by reverse-engineering and it was easier than to implement any intermediate low level standard layer, BUT Django releases a new version every 8 months that requires to debug, rethink and refactor approximately one third of our code, because the ORM is substantionally improved regularly. The SQL and SOQL references are changing much slower. I see that the current Salesforce backend require more maintenance than all standard backends together.

A naive temporary fix is probably still possible with many restrictions, but I prefer to write a new stable layer that will imitate the basic principles of the DB API. The current SOQL compiler is the first and successful step. The soap and bulk api is importand not only due to speed but also to prevent governor limits for APIs, otherwise you can only click on SFDC web for that company for 24 hours without using any API. I want to be ready with this before [Django 1.10 -.Schedule: July 2016](https://code.djangoproject.com/wiki/Version1.10Roadmap#schedule). Versions 1.8 - 1.10 will be supported.

**Comment by hynekcer** [comment]
(I know that this issue is useful. I would think about a simple temporary solution to bridge the gap until the code will be refactored and well tested again.)

You wrote that this software is useful for you. I tried to contact you last week, but without success. Please can you express how much you appreciate its benefits for you? (personal email) It is so that I publish everything what seems sustainable for future and generally useful. I'm free from any business obligations, but now I realize that with no real material evaluation of this work I can not continue for long time, especially if I expect that people probably should use it commercially. I use it for a few [NGO](https://en.wikipedia.org/wiki/Non-governmental_organization)s that I want to support hard not by money, but my work. You understand that it's me too generous in many directions. Currently it would be more profitable to wash dishes half-time :-)

**Comment by hynekcer** [comment]
Almost all regressions by Django 1.8 are fixed fixed in my development branches thanks to removing support for Django 1.7 and subsequently a more clear code. A complete solution for values() and values_list() is still complicated. Fortunately, your case is simple enough to be fixed together with the release for Django 1.10, though not fixed yet.

**Comment by hynekcer** [comment]
A solution already exists even for Django 1.10, see #160. The new solution doesn't require to be patched so frequently for new Django versions. It required refactoring of hundreds lines. It takes some time to test it enough and think forward, before committing to master, so as the future development can be straight without regression.

**Comment by hynekcer** [comment]
Fixed by PR #222
